packages/node/src/indexer/worker/block-dispatcher.service.ts

```diff
@@ -0,0 +1,383 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import assert from 'assert';
+import os from 'os';
+import path from 'path';
+import { Injectable, OnApplicationShutdown } from '@nestjs/common';
+import { RuntimeVersion } from '@polkadot/types/interfaces';
+import { SubstrateBlock } from '@subql/types';
+import { EventEmitter2 } from 'eventemitter2';
+import { last } from 'lodash';
+import { NodeConfig } from '../../configure/NodeConfig';
+import { AutoQueue, Queue } from '../../utils/autoQueue';
+import { getLogger } from '../../utils/logger';
+import { fetchBlocksBatches } from '../../utils/substrate';
+import { ApiService } from '../api.service';
+import { IndexerEvent } from '../events';
+import { IndexerManager } from '../indexer.manager';
+import {
+  FetchBlock,
+  ProcessBlock,
+  InitWorker,
+  NumFetchedBlocks,
+  NumFetchingBlocks,
+  SetCurrentRuntimeVersion,
+} from './worker';
+import { Worker } from './worker.builder';
+
+type IIndexerWorker = {
+  processBlock: ProcessBlock;
+  fetchBlock: FetchBlock;
+  numFetchedBlocks: NumFetchedBlocks;
+  numFetchingBlocks: NumFetchingBlocks;
+  setCurrentRuntimeVersion: SetCurrentRuntimeVersion;
+};
+
+type IInitIndexerWorker = IIndexerWorker & {
+  initWorker: InitWorker;
+};
+
+type IndexerWorker = IIndexerWorker & {
+  terminate: () => Promise<number>;
+};
+
+async function createIndexerWorker(): Promise<IndexerWorker> {
+  const indexerWorker = Worker.create<IInitIndexerWorker>(
+    path.resolve(__dirname, '../../../dist/indexer/worker/worker.js'),
+    [
+      'initWorker',
+      'processBlock',
+      'fetchBlock',
+      'numFetchedBlocks',
+      'numFetchingBlocks',
+      'setCurrentRuntimeVersion',
+    ],
+  );
+
+  await indexerWorker.initWorker();
+
+  return indexerWorker;
+}
+
+type GetRuntimeVersion = (block: SubstrateBlock) => Promise<RuntimeVersion>;
+
+function getMaxWorkers(numWorkers?: number): number {
+  const maxCPUs = os.cpus().length - 1;
```

Comment:
```suggestion
  const maxCPUs = os.cpus().length;
```
Main thread doesn't do a lot things. It doesn't make sense if I need 3 cpu so I can have two workers.